pkg/virtctl/credentials/add-key/add-key.go

```diff
@@ -0,0 +1,297 @@
+	cmd.Flags().BoolVar(&a.UpdateSecret, updateSecretFlag, false, "Add the SSH key to an existing secret. Use --force option, if the secret does not have owner reference pointing to the VM.")
+	cmd.MarkFlagsMutuallyExclusive(createSecretFlag, updateSecretFlag)
+
+	cmd.Flags().BoolVar(&a.Force, "force", false, "Force update of secret, even if it's not owned by the VM.")
+}
+
+func runAddKeyCommand(clientConfig clientcmd.ClientConfig, cmdFlags *addSshKeyFlags, cmd *cobra.Command, args []string) error {
+	vmName := args[0]
+	vmNamespace, _, err := clientConfig.Namespace()
+	if err != nil {
+		return fmt.Errorf("error getting namespace: %w", err)
+	}
+
+	// Reading the key before accessing cluster
+	sshKey, err := common.GetSshKey(&cmdFlags.SshCommandFlags)
+	if err != nil {
+		return fmt.Errorf("error getting ssh key: %w", err)
+	}
+
+	cli, err := kubecli.GetKubevirtClientFromClientConfig(clientConfig)
+	if err != nil {
+		return fmt.Errorf("error getting kubevirt client: %w", err)
+	}
+
+	vm, err := cli.VirtualMachine(vmNamespace).Get(cmd.Context(), vmName, &metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("error getting virtual machine: %w", err)
+	}
+
+	if shouldCreateNewSecret(cmdFlags, vm) {
+		return addSecretWithSshKey(cmd, cli, cmdFlags, vm, sshKey)
+	} else {
+		return updateSecretWithSshKey(cmd, cli, cmdFlags, vm, sshKey)
+	}
+}
+
+func addSecretWithSshKey(cmd *cobra.Command, cli kubecli.KubevirtClient, cmdFlags *addSshKeyFlags, vm *v1.VirtualMachine, sshKey string) (err error) {
+	if !cmdFlags.Force {
+		// Only create a secret if VM is not running.
+		_, err := cli.VirtualMachineInstance(vm.Namespace).Get(cmd.Context(), vm.Name, &metav1.GetOptions{})
+		if err == nil {
+			return fmt.Errorf("virtual machine %s is running", vm.Name)
+		}
+		if !errors.IsNotFound(err) {
+			return fmt.Errorf("error when getting virtual machine instance: %w", err)
+		}
+	}
+
+	secret := newSecretWithKey(vm, sshKey)
+	secret, err = cli.CoreV1().Secrets(vm.Namespace).Create(cmd.Context(), secret, metav1.CreateOptions{})
+	if err != nil {
+		return fmt.Errorf("error creating secret: %w", err)
+	}
+
+	vmPatched := false
+	defer func() {
+		// If VM patch failed, remove the secret
+		if vmPatched {
+			return
+		}
+		deleteErr := cli.CoreV1().Secrets(vm.Namespace).Delete(cmd.Context(), secret.Name, metav1.DeleteOptions{})
+		if deleteErr != nil && err == nil {
+			// Return the deletion error from the enclosing function
+			err = fmt.Errorf("error deleting secret: %w", deleteErr)
+		}
+	}()
+
+	accessCredential := newAccessCredential(secret.Name, cmdFlags.User)
+
+	patchData := patchToAddAccessCredential(accessCredential)
+	_, err = cli.VirtualMachine(vm.Namespace).Patch(cmd.Context(), vm.Name, types.JSONPatchType, patchData, &metav1.PatchOptions{})
+	if err != nil {
+		// TODO: Check error type and only execute second patch for some errors
+		// Try adding the array
+		patchData := patchToAddAccessCredentialArray(accessCredential)
```

Comment:
When is this necessary?

Reply:
If the VM has no credentials, the slice is `nil`, and patch will fail, because path `/spec/template/spec/accessCredentials/-` does not exist. So this second patch is used to add the array.